### PR TITLE
fix: avoid scanning through all local file headers when opening an archive

### LINKED ISSUE #280: zip 2.2.2 scans for large parts of the file while opening a ZipArchive
**Describe the bug**

(I know about recently closed #231)

I tried 2.2.2 on a 100+GB zip file on a network storage, and `get_metadata` still scans large parts of the file because of this validation in `central_header_to_zip_file`:

https://github.com/zip-rs/zip2/blob/7c20fa30016301909bf2ade203cb4841b7776154/src/read.rs#L1071-L1077

There might be a reason (an invariant) for this check to run already when opening the archive, but for random access of big files (especially over the network), it would be good if those checks would be deferred until the file is accessed.

Full stack trace:

```
   3: zip::read::find_data_start
             at /home/johannes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zip-2.2.2/src/read.rs:352:5
   4: zip::read::central_header_to_zip_file
             at /home/johannes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zip-2.2.2/src/read.rs:1071:22
   5: zip::read::<impl zip::read::zip_archive::ZipArchive<R>>::read_central_header
             at /home/johannes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zip-2.2.2/src/read.rs:654:24
   6: zip::read::<impl zip::read::zip_archive::ZipArchive<R>>::get_metadata::{{closure}}
             at /home/johannes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zip-2.2.2/src/read.rs:616:34
   7: core::result::Result<T,E>::and_then
             at /home/johannes/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1348:22
   8: zip::read::<impl zip::read::zip_archive::ZipArchive<R>>::get_metadata
             at /home/johannes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zip-2.2.2/src/read.rs:615:30
   9: zip::read::<impl zip::read::zip_archive::ZipArchive<R>>::with_config
             at /home/johannes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zip-2.2.2/src/read.rs:714:22
  10: zip::read::<impl zip::read::zip_archive::ZipArchive<R>>::new
             at /home/johannes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zip-2.2.2/src/read.rs:707:9
```

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #280

The idea is to make `ZipFileData.data_start` be calculated lazy to avoid accessing all local file headers already when opening a file. This required a change to the signature of `ZipFileData.data_start()` (which seems to be non-public after all).

### CODE REVIEW

File: src/read.rs [hunk 1]
```diff
@@ -1068,14 +1065,6 @@ pub(crate) fn central_header_to_zip_file<R: Read + Seek>(
         ));
     }
 
-    let data_start = find_data_start(&file, reader)?;
-
-    if data_start > central_directory.directory_start {
```
Comment (Pr0methean) [inline src/read.rs:1073]: Where has this check gone?

**Review by Pr0methean — CHANGES_REQUESTED** [review]

**Review by Pr0methean — CHANGES_REQUESTED** [review]
Sounds good in principle, but I don't like the idea of totally removing the validation when we could just defer it.

File: src/read.rs (hunk 1 above)
Comment (jrudolph) [reply src/read.rs:1259]: The question is what the ultimate purpose of this check is. It looks a bit like a conservative check of some specification requirement.

But is it require for correctness / ruling out weird edge cases? After all no tests fail after removing this check.

In some way, being able to do this check runs counter to the idea of this PR of not having to scan through the file for random access. I.e. even if we defer, e.g. to the below `ZipFileData.data_start` or move the check into `find_data_start`, the random access use case I have in mind here will never execute the check.

Comment (Pr0methean) [reply src/read.rs:1259]: IIRC it does relate to edge cases, such as when archives are concatenated or nested. Even if the spec is ambiguous in those cases about which one to extract, which I'm pretty sure is true for concatenation when the second one is under 64KiB, we should still consistently choose one or the other.

**Review by Pr0methean — APPROVED** [review]
After this is merged I'll look at where the data-start-after-header-start check might be safely merged, to ensure we're not reading an old central directory that's been superseded.

**Comment by Lynnesbian** [comment]
@Pr0methean Just a suggestion: If you can't find a way to add the check back without incurring a performance penalty, maybe it could be made optional with a parameter on the [`Config`](https://docs.rs/zip/latest/zip/read/struct.Config.html) struct?
